File: Matlab/tester.m
% tester
%% Load data for work
load('dataTr.mat')
%% test of oneDClass and Fisher
% Index of positive (bad) outcome
ind = outcome == 1;
x = data(ind, :);
y = data(~ind, :);
accs = ["BA", "Acc", "f1", "NPV", "PPV", "TPR", "TNR"];
for ac = accs
    [bestT, bestErr, dir] = fisher(x, y, {"LFD score for Binary GOSE outcome", "Bad", "Good"}, ac);
    saveFigures(['figures/', char(ac), '.png']);
end

%% Test specDir
[bestT, bestErr, dir] = fisher(x, y, {"LFD score for Binary GOSE outcome", "Bad", "Good"}, 'ba');
for k = 1:5
    ddir = rand(size(x, 2), 1);
    [bestT, bestErr, dir1] = specDir(x, y, ddir, {"Random score for Binary GOSE outcome", "Bad", "Good"}, 'ba');
    saveFigures(['figures/Random_', num2str(k), '.png']);
end

%% Comparison of direction usage
[bestT, bestErr, dir] = fisher(x, y, {"LFD score for Binary GOSE outcome", "Bad", "Good"}, 'ba');
[bestT, bestErr, dir1] = specDir(x, y, -dir, {"Random score for Binary GOSE outcome", "Bad", "Good"}, 'ba');

%% Test of function max search methods.
nRep = 10000;
Da = [0.9, 0.9, 0.9, 0.5, 0.5, 0.5];
Ns = [10, 100, 1000, 10, 100, 1000];
% Grid search
for K = 1:6
    tic;
    for k=1:nRep
        N=10000;
        Delta=Da(K);
        % interval in which we want to find inf / sup of rho and psi
        eps = linspace(0,Delta,N);
        % probability threshold
        N_sample = Ns(K);

        buff=(Delta - eps) .* (1 - 2 * exp(-2 * eps .^2 * N_sample));
        rho = max(max(buff), 0);
    end
    a = toc;
    fprintf("Delta %f Ns %d rho %f time %f\n", Da(K), Ns(K), rho, a);
end
% Dichotomy like
% for K = 1:6
%     tic;
%     Delta=Da(K);
%     N_sample = Ns(K);
%     for k=1:nRep
%         [~, rho] = fminbnd(@(eps) -(Delta - eps) * (1 - 2 * exp(-2 * eps ^2 * N_sample)), 0, Delta);
%     end
%     a = toc;
%     fprintf("Delta %f Ns %d rho %f time %f\n", Da(K), Ns(K), rho, a);
% end



%% test of final function
load('dataSplit.mat');

% Form model for training set
[bestT, bestErr, dir] = fisher(xTr, yTr, [], 'ba');

% Assess this model for test set
xSc = xTe * dir;
ySc = yTe * dir;

% Form labels and predictions
labX = zeros(length(xSc), 1);
labY = ones(length(ySc), 1);
predX = labX;
predY = labY;
% Form sets for estimation
ind = xSc >= bestT;
predX(ind) = 1;
ind = ySc < bestT;
predY(ind) = 0;

data = [xTe; yTe];
lab = [labX; labY];
pred = [predX; predY];
%
lab = lab+1;
pred = pred+1;
%% Test itself
mdl = modelCreator(data, lab, pred);

%% Test of estimate
[gen, det] = estimate(mdl);

File: Python/tester.py
# -*- coding: utf-8 -*-
"""
Created on Thu Mar  7 19:09:28 2024

@author: em322
"""

import numpy as np
from guarantee import *
from random import sample
from math import floor

# Load data and prepare datasets
with open('QSAR.csv') as file:
    content = file.readlines()
# Remove headers
content = content[1:]
# Create tables for reading
data = np.zeros((len(content), 41))
outcome = np.zeros(len(content))
for k in range(len(content)):
    tmp = content[k].split(",")
    ttt = [float(i) for i in tmp]
    data[k, :] = np.asarray(ttt[0:-1])
    outcome[k] = ttt[-1]

# Test of oneDClass
# Index of positive (bad) outcome
ind = outcome == 1;
x = data[ind, :]
y = data[~ind, :]
accs = ["BA", "Acc", "f1", "NPV", "PPV", "TPR", "TNR"]
for ac in accs:
    [bestT, bestErr, dir] = fisher(x, y, ["LFD score for Binary GOSE outcome", "Bad", "Good"], ac)

[bestT, bestErr1, dir] = fisher(x, y, acc='ba')

#Test of randomly selected. It is infinite loop to find good direction. 
# Please decomment following rows to test, if you want
# bestErr = float('inf')
# bestDir = None
# siz = x.shape[1]
# while True:
#     dirs = np.random.rand(siz)
#     [t, err, dirs] = specDir(x, y, dirs, acc='BA')
#     if err < bestErr:
#         bestDir = dirs
#         bestErr = err
#         print(err)

# Split data into test and training set
ind = sample(range(x.shape[0]), floor(0.2 * x.shape[0]))
xTe = x[ind, :];
xTr = np.delete(x, ind, 0)
ind = sample(range(y.shape[0]), floor(0.2 * y.shape[0]))
yTe = y[ind, :];
yTr = np.delete(y, ind, 0)

# Form model for training set
[bestT, bestErr, direct] = fisher(xTr, yTr, None, 'ba');

# Assess this model for test set
xSc = np.matmul(xTe, direct);
ySc = np.matmul(yTe, direct)

# Form labels and predictions
labX = np.zeros(xSc.shape[0])
labY = np.ones(ySc.shape[0])
predX = np.zeros(xSc.shape[0])
predY = np.ones(ySc.shape[0])

# Form sets for estimation
ind = xSc >= bestT;
predX[ind] = 1
ind = ySc < bestT;
predY[ind] = 0

data = np.concatenate((xTe, yTe))
lab = np.concatenate((labX, labY))
pred = np.concatenate((predX, predY))
# According to documentations classes must be numerated from 1.
lab = lab+1;
pred = pred+1;

# Test of rejectAcceptModel creation
mdl = RejectAcceptModel(data, lab, pred) 
